nemo/collections/llm/modelopt/model_utils.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utility functions for loading models with modelopt layer spec."""
+
+import lightning.pytorch as L
+import torch
+from megatron.core.dist_checkpointing.validation import StrictHandling
+
+from nemo import lightning as nl
+from nemo.collections import llm
+from nemo.collections.llm.inference.base import _setup_trainer_and_restore_model
+from nemo.collections.nlp.modules.common.tokenizer_utils import get_tokenizer
+from nemo.lightning.ckpt_utils import ckpt_to_context_subdir
+from nemo.lightning.io.pl import ckpt_to_weights_subdir
+from nemo.utils import logging
+
+__all__ = ["set_modelopt_spec_if_exists_in_ckpt", "setup_trainer_and_restore_model_with_modelopt_spec"]
+
+
+def _set_gpt_modelopt_spec(model_cfg: llm.GPTConfig) -> llm.GPTConfig:
+    """Set model.config.transformer_layer_spec to modelopt spec."""
+    logging.info("Setting model.config.transformer_layer_spec to gpt_modelopt_spec")
+    assert isinstance(model_cfg, llm.GPTConfig), "model_cfg must be a GPTConfig"
+    try:
+        from functools import partial
+
+        from megatron.core.post_training.modelopt.gpt.model_specs import get_gpt_modelopt_spec
+
+        modelopt_spec = partial(get_gpt_modelopt_spec, remap_te_layernorm=True)
+    except ImportError:
+        # Older spec: Will be deprecated, doesnt support DeepSeek
+        from megatron.core.inference.modelopt_support.gpt.model_specs import get_gpt_layer_modelopt_spec
+
+        modelopt_spec = get_gpt_layer_modelopt_spec(num_experts=model_cfg.num_moe_experts, remap_te_layernorm=True)
+    model_cfg.transformer_layer_spec = modelopt_spec
+    return model_cfg
+
+
+def set_modelopt_spec_if_exists_in_ckpt(model: L.LightningModule, path: str) -> bool:
```

Comment:
new utility function to allow automatically using modelopt spec if restoring a ckpt that has `modelopt_state` in it (i.e. quantized ckpt)